### ISSUE TITLE
Does NNI support HPO on PyTorch DDP training?

### ISSUE THREAD

**Issue by shiyf369** [microsoft/nni#3510]
Hi, I'd like to use automatic Hyper Parameter Optimization of NNI on PyTorch DDP distributed training.  
Python API:  from torch.nn.parallel import DistributedDataParallel

According to NNI HPO example, I need to execute "nnictl create --config config.yml" to start HPO.
But PyTorch DDP need to execute training command on each distributed node.
Does NNI support HPO on PyTorch DDP training? Is there an example I can refer to ?
Thank you very much.

**Comment by QuanluZhang** [comment]
hi @shiyf369 , yes, NNI supports HPO on Pytorch DDP training. An example is here: https://github.com/microsoft/nni/tree/master/examples/trials/mnist-distributed-pytorch

We support distributed training on training services such as kubeflow, AML, OpenPAI. in these training services, you can specify workers in yaml file.

**Comment by shiyf369** [comment]
Thank you for your reply.

When I test HPO on PyTorch DDP, I found:
1. NNI doesn't start process of the remote node for PyTorch DDP distributed training, the master node has been waiting for the remote process.
2. nni.report_intermediate_result() can't report result to webserver, the intermediate result can't be shown on webUI.
3. command: “python3 -m torch.distributed.launch --nnodes=2 --node_rank=0 --nproc_per_node 1”, when I set nproc_per_node as 1, the program is stuck. But when I set nproc_per_node as 2, the program can run normally.

**Comment by QuanluZhang** [comment]
@shiyf369 , you are using kubeflow training service, and have set up kubeflow cluster?

**Comment by shiyf369** [comment]
no, I use remote mode.  trainingServicePlatform: remote

this is my config file:

**Comment by shiyf369** [comment]
no, I use remote mode.  trainingServicePlatform: remote

this is my config file:

**Comment by shiyf369** [comment]
`authorName: default
experimentName: My_example_mnist_distributed_pytorch
trialConcurrency: 1
maxExecDuration: 1h
maxTrialNum: 10
#choice: local, remote, pai, kubeflow
trainingServicePlatform: remote
searchSpacePath: search_space.json
#choice: true, false
useAnnotation: false
nniManagerIp: 189.105.115.145
tuner:
  #choice: TPE, Random, Anneal, Evolution
  builtinTunerName: TPE
  classArgs:
    #choice: maximize, minimize
    optimize_mode: minimize
trial:
  command: CUDA_VISIBLE_DEVICES="2,3" python3 -m torch.distributed.launch --nnodes=2 --node_rank=0 --nproc_per_node 2 --master_addr 189.105.115.145 --master_port 29506 dist_mnist.py
  codeDir: .
  gpuNum: 2

machineList:
  - ip: 189.105.115.145
    port: 22
    username: root
    passwd: admin123
    gpuIndices: 2,3
  - ip: 189.105.115.146
    port: 22
    username: root
    passwd: admin123
    gpuIndices: 2,3

`

**Comment by shiyf369** [comment]
this is my training script:

`import nni
import logging

import argparse
from tqdm import tqdm
import torch
import torchvision
import torch.nn as nn
import torch.nn.functional as F

import torch.distributed as dist
from torch.nn.parallel import DistributedDataParallel as DDP

logger = logging.getLogger('nni_pytorch_dist')

class ToyModel(nn.Module):
    def __init__(self):
        super(ToyModel, self).__init__()
        self.conv1 = nn.Conv2d(3, 6, 5)
        self.pool = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(6, 16, 5)
        self.fc1 = nn.Linear(16 * 5 * 5, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)

    def forward(self, x):
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = x.view(-1, 16 * 5 * 5)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x


def get_dataset():
    transform = torchvision.transforms.Compose([
        torchvision.transforms.ToTensor(),
        torchvision.transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
    ])
    my_trainset = torchvision.datasets.CIFAR10(root='./data', train=True,
                                               download=True, transform=transform)
    train_sampler = torch.utils.data.distributed.DistributedSampler(my_trainset)
    trainloader = torch.utils.data.DataLoader(my_trainset,
                                              batch_size=16, num_workers=2, sampler=train_sampler)
    return trainloader


def generate_default_params():
    '''
    Generate default parameters for mnist network.
    '''
    params = {
        'learning_rate': 0.01,
        'momentum': 0.5}
    return params

if __name__ == "__main__":
    RCV_PARAMS = nni.get_next_parameter()
    logger.debug(RCV_PARAMS)
    params = generate_default_params()
    params.update(RCV_PARAMS)

    parser = argparse.ArgumentParser()
    parser.add_argument("--local_rank", default=-1, type=int)
    FLAGS = parser.parse_args()
    local_rank = FLAGS.local_rank

    torch.cuda.set_device(local_rank)

    print("Start to init ddp init_process_group ...")
    dist.init_process_group(backend='nccl')
    print("Finish ddp init_process_group.")

    trainloader = get_dataset()

    model = ToyModel().to(local_rank)
    ckpt_path = None
    if dist.get_rank() == 0 and ckpt_path is not None:
        model.load_state_dict(torch.load(ckpt_path))
    # DDP: build DDP model
    model = DDP(model, device_ids=[local_rank], output_device=local_rank)
    optimizer = torch.optim.SGD(model.parameters(), lr=params['learning_rate'], momentum=params['momentum'])
    loss_func = nn.CrossEntropyLoss().to(local_rank)

    model.train()
    iterator = tqdm(range(100))
    rank = dist.get_rank()
    print("Start to train model...")
    total_loss = 0.0
    for epoch in iterator:
        print("Start new epoch: {}".format(epoch))
        epoch_loss = 0.0
        trainloader.sampler.set_epoch(epoch)
        for data, label in trainloader:
            data, label = data.to(local_rank), label.to(local_rank)
            optimizer.zero_grad()
            prediction = model(data)
            loss = loss_func(prediction, label)
            epoch_loss += loss.item()
            loss.backward()
            iterator.desc = "loss = %0.3f" % loss
            optimizer.step()

        if rank == 0:
            print("nni.report_intermediate_result: {}".format(epoch_loss / len(trainloader.dataset)))
            nni.report_intermediate_result(epoch_loss / len(trainloader.dataset))

        total_loss += (epoch_loss / len(trainloader.dataset))

    total_loss /= 100
    logger.debug('Final loss: {}'.format(total_loss))
    if rank == 0:
        print("nni.report_final_result: {}".format(total_loss))
        nni.report_final_result(total_loss)
`

**Comment by shiyf369** [comment]
@QuanluZhang  Doesn't NNI support remote mode for PyTorch DDP training?

**Comment by QuanluZhang** [comment]
@shiyf369 , no, but you have two options to workaround.

let's say one DDP trial runs on three remote node. and you have 3 remote nodes in total.

option 1, set trialConcurrency 3. each worker detect whether it runs on No.1 machine, if it is, it is master, if not, it is worker. In this case, one DDP trial runs as three NNI trials. The drawback of this workaround is that configurations on the two workers are wasted.

option2, set trialConcurrency 1, and only configure one remote node in your experiment yaml config file. trial command should be responsible for launching workers on the other two nodes. The drawback of this workaround is that you need to write the script to launch workers remotely.

**Comment by shiyf369** [comment]
Got it. Thank you.
